### PR TITLE
APQ: Add cache-control header on PERSISTED_QUERY_NOT_FOUND

### LINKED ISSUE #2502: APQ: Add cache-control header on `PERSISTED_QUERY_NOT_FOUND`
The router doesn't add any cache control header to the APQ `PERSISTED_QUERY_NOT_FOUND`, which means CDNs could cache the temporary error.

as @lleadbet found out, [AS4 does the right thing](https://github.com/apollographql/apollo-server/blob/d7e9b97595b063f1e796ec4449850a16d19e8b18/packages/server/src/internalErrorClasses.ts#L51-L85) and sets `cache-control` to `private, no-cache, must-revalidate`.

This issue might be the right opportunity to have a more principled approach to error handling and cache headers.

### PR DESCRIPTION
**Pull request by o0Ignition0o**
Fixes #2502 

The router now sends `cache-control: private, no-cache, must-revalidate` when a persisted query hash was not found.

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/cache/mod.rs [hunk 1]
```diff
@@ -113,7 +113,11 @@ where
     }
 
     pub(crate) async fn insert(&self, key: K, value: V) {
-        self.storage.insert(key, value.clone()).await;
+        // Let's see if we can help waiters make progress.
+        if let Some(sender) = self.wait_map.lock().await.remove(&key) {
+            let _ = sender.send(value.clone());
+        }
```
Comment (o0Ignition0o) [inline apollo-router/src/cache/mod.rs:119]: cc @Geal @garypen because I honestly don't understand all the intricacies of the cache, all I know is the test used to fail, and not it passes.

Comment (garypen) [reply apollo-router/src/cache/mod.rs:119]: I think the problem you are resolving is that there may be waiters who are not notified that a value has been inserted. The bit I don't understand is that, if there are waiters, then something should be "getting" the cached value and that should resolve at some point.

Logically, it's probably not safe to insert() a value to the cache if there are waiters for that key, so I think that aspect of your change is correct, but I don't like the fact that waiters should be resolved at some point and maybe insert should be ordered behind the existing get?

I think it needs input from @Geal .

Comment (o0Ignition0o) [reply apollo-router/src/cache/mod.rs]: right, we paired on this and it turns out to be a corner case we hit because the test runs in the single threaded flavor. the actual observable behavior is that if it ever happens, one(1) client will need to send the apq and the query once more, which is totally ok.

Moreover this allows us to /not/ send the value like that. the core issue stems from the fact that the cache assumes the first person who tries to `get` is going to `insert`, which doesn't apply to APQ.

File: apollo-router/src/services/layers/apq.rs [hunk 2]
```diff
@@ -79,6 +83,7 @@ async fn apq_request(
                 request.supergraph_request.body_mut().query = Some(cached_query);
                 Ok(request)
             } else {
+                let _ = request.context.insert("persisted_query_miss", true);
```
Comment (bnjjj) [inline apollo-router/src/services/layers/apq.rs:86]: Is it still used ? I can't find the corresponding get. But if it's still relevant then please use a constant with some namespacing like `apq::persisted_query_miss`

**Review by BrynCooke — APPROVED** [review]
